analyzers/src/SonarAnalyzer.Common/SyntaxTrackers/BaseTypeTracker.cs

```diff
@@ -114,5 +64,28 @@ internal BaseClassCondition MatchSubclassesOf(params KnownType[] types)
                 return false;
             };
         }
+
+        protected override BaseContext IsTracked(SyntaxNode expression, SemanticModel semanticModel, BaseClassCondition[] conditions, out Location location)
+        {
+            location = Location.None;
+
+            var baseTypeList = GetBaseTypeNodes(expression);
+            if (baseTypeList == null || !baseTypeList.Any())
+            {
+                return null;
+            }
+
+            var context = new BaseTypeContext(expression, baseTypeList, semanticModel);
+
+            // We can't pass the issueLocation to the lambda directly so we need a temporary variable
+            Location locationToReport = null;
+            if (conditions.All(c => c(context, out locationToReport)))
```

Comment:
This is really ugly old design. The location magic should happen inside that specific condition, directly to it's context.

That makes all the condition delegates the same. So there should be just a single generic condition delegate in the base type. And the `conditions.All` part could be moved to base class, as mentioned in another comment.